File: src/py/test/test_ast_utils/test_utils.py
import unittest
import sys
sys.path.insert(0, 'src/py') # hack for now

import ast
from ast_utils.utils import *
from ast_utils.preprocess import *
from ast_utils.node_finder import NodeFinder

class TestUtils(unittest.TestCase):
    def test_get_name(self):
        source_code = "x = 1"
        node = ast.parse(source_code).body[0]
        self.assertEquals(get_name(node.targets[0]).id, "x")

        source_code = "x[i] = 1"
        node = ast.parse(source_code).body[0]
        self.assertEquals(get_name(node.targets[0]).id, "x")

        source_code = "x: int = 1" # ast.AnnAssign
        node = ast.parse(source_code).body[0]
        self.assertEquals(get_name(node.target).id, "x")

    def test_get_call_name(self):
        source_code = "func(x)"
        node = ast.parse(source_code).body[0].value
        self.assertEquals(get_call_name(node), "func")

        source_code = "module.func(x)"
        node = ast.parse(source_code).body[0].value
        self.assertEquals(get_call_name(node), "func")


    def test_block_and_is_different_branch(self):
        source_code = """
if test1:
    A
    B
else:
    C
    if test2:
        D
    else:
        E
    if test3:
        F
    if test4:
        pass
        """
        parsed_ast = ast.parse(source_code)
        line_offsets = get_line_offsets_for_str(source_code)
        syntax_tree = preprocess_syntaxtree(parsed_ast, source_code, line_offsets, 0)

        node = syntax_tree.root_node.body[0]

        A = node.body[0].value
        B = node.body[1].value
        C = node.orelse[0].value
        D = node.orelse[1].body[0].value
        E = node.orelse[1].orelse[0].value
        F = node.orelse[2].body[0].value
        # [ast.dump(n) for n in [A,B,C,D,E,F]]
        
        self.assertTrue(isinstance(node.body, Block)) # [A,B]
        self.assertTrue(len(node.body) == 2)
        self.assertTrue(node.body[0].value == A and node.body[1].value == B)
        self.assertTrue(is_descendant(node.body[0], A))
        self.assertTrue(isinstance(node.orelse, Block)) # [C, if ...]
        self.assertTrue(isinstance(node.orelse[1].body, Block)) # D
        self.assertTrue(isinstance(node.orelse[1].orelse, Block)) # E
        self.assertTrue(isinstance(node.orelse[3].body, Block)) # pass
        self.assertTrue(not hasattr(node.orelse[3], "orelse"))
        
        self.assertTrue(not is_in_different_branch(A,B))
        self.assertTrue(is_in_different_branch(A,C) and is_in_different_branch(A,D) and is_in_different_branch(A,E) and is_in_different_branch(A,F))
        self.assertTrue(not is_in_different_branch(C,F))
        self.assertTrue(is_in_different_branch(D,E))
        self.assertTrue(not is_in_different_branch(D,F))

#     def test_get_subnode_for_range(self):
#         source_code = """
# 1
# x = 1
# def test(a, b: int, c=1, d: int=1):
#     if a == 1:
#         y = 2
#     else:
#         x = 3
    
#     for i in range(b,f):
#         if i > 10:
#             break

#     while c > 2:
#         c = c - 1
# """
#         parsed_ast = ast.parse(source_code)
#         line_offsets = get_line_offsets_for_str(source_code)
#         _ = preprocess_syntaxtree(parsed_ast, source_code, line_offsets, 0)

#         def nodefinder_map(x):
#             print(ast.dump(x))
#             a, b = get_range(x)
#             print(a,b)
#             print(source_code[a:b])
#             return get_subnode_for_range(syntax_tree, a, b) == x

#         nodefinder = NodeFinder(lambda x: True, nodefinder_map, visit_matched_nodes=True)
#         nodefinder.visit(syntax_tree)
#         self.assertTrue(all(nodefinder.result))

File: src/jl/test/ast/test_utils.jl

@testset "utils" begin
    
    source_code = "function func(x::Int) end"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_function_name(node) == :func

    source_code = "function func(x::Int)::Int end"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_function_name(node) == :func

    source_code = "function Module.func(x::Int) end"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_function_name(node) == :func

    source_code = "function Module.func(x::Int)::Int end"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_function_name(node) == :func


    source_code = "func(x)"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_call_name(node) == :func

    source_code = "Module.func(x)"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_call_name(node) == :func

    source_code = "Vector{Int}(undef, 10)"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_call_name(node) == Symbol("Vector{Int}")

    source_code = "function test(a, b::Int, c=1, d::Int=1; e, f::Int, g=1, h::Int=1) end"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_parameter_names_of_function(node) == [:a, :b, :c, :d, :e, :f, :g, :h]

    source_code = "function test(a, b::Int, c=1, d::Int=1; e, f::Int, g=1, h::Int=1)::Int end"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_parameter_names_of_function(node) == [:a, :b, :c, :d, :e, :f, :g, :h]


    source_code = "x = 1"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_assignment_identifier(node).val == :x

    source_code = "x::Int = 1"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_assignment_identifier(node).val == :x

    source_code = "x[i] = 1"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_assignment_identifier(node).val == :x

    source_code = "x::Vector{Int} = Int[]"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test get_assignment_identifier(node).val == :x

    source_code = "
    1
    x = 1
    function test(a, b::Int, c=1, d::Int=1; e, f::Int, g=1, h::Int=1)::Int
        if a == 1
            y = 2
        else
            x = 3
        end
        for i = b:f
            if i > 10
                break
            end
        end
        while c > 2
            c = c - 1
        end
    end
    "
    syntaxtree = JuliaSyntax.parseall(SyntaxNode, source_code)
    nodefinder = NodeFinder{Bool}(x -> true, x -> get_subnode_for_range(syntaxtree, range(x)) == x; visit_matched_nodes=true)
    result = visit(nodefinder, syntaxtree)
    @test all(result)


    source_code = "x"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1]
    @test kind(node) == K"Identifier" && is_variable_identifier(node)

    source_code = ":x"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1,1]
    @test kind(node) == K"Identifier" && !is_variable_identifier(node)

    source_code = "Main.x"
    node = JuliaSyntax.parseall(SyntaxNode, source_code)[1,2,1]
    @test kind(node) == K"Identifier" && is_variable_identifier(node)
    
    source_code = """
    if test1
        A
        B
    else
        C
        if test2
            D
        else
            E
        end
        if test3
            F
        end
    end
    """
    syntaxtree = JuliaSyntax.parseall(SyntaxNode, source_code)

    A = syntaxtree[1,2,1]
    B = syntaxtree[1,2,2]
    C = syntaxtree[1,3,1]
    D = syntaxtree[1,3,2,2,1]
    E = syntaxtree[1,3,2,3,1]
    F = syntaxtree[1,3,3,2,1]

    @test !is_in_different_branch(A,B)
    @test is_in_different_branch(A,C) && is_in_different_branch(A,D) && is_in_different_branch(A,E) && is_in_different_branch(A,F)
    @test !is_in_different_branch(C,F)
    @test is_in_different_branch(D,E)
    @test !is_in_different_branch(D,F)
end
